[WIP] Feature/0133 ppcs

bob-carpenter opened the pull request:
Oops.  Forgot to complete the cross-validation section!  I'll remove the WIP tag when that's done.

#### Submission Checklist

- [x] Builds locally 
- [x] Declare copyright holder and open-source license: see below

#### Summary

Adds a new part to the user's guide on meta-algorithms (one that require multiple Stan fits).  Chapters are:

* posterior predictive checks

* simulation-based calibration

* cross-validation

* bootstrap (yes, classical bootstrap, with a quick mention of bagging and Bayesian bootstrap)

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

avehtari commented:
I'm happy to review this, when it's ready for review

mitzimorris commented:
img files not checked in:

- `img/sho-ode-trajectory.png`
- `img/change-point-posterior.png`
- `img/s-discrete-posterior.png`

bob-carpenter commented:
@avehtari --- this is ready for review.

bob-carpenter commented:
Comments sent by email from @andrewgelman:

> I recommend using h() rather than f() for an arbitrary function.  The trouble is that sometimes people use f for the probability density function.  So the f on page 284 could confuse people.  You could use g, but I usually use h, because sometimes I use g for an approximation to p when doing importance sampling etc.

OK.  I usually avoid `pi` as a variable name for the same reason.  Of course, the same people who write `f` for densities will also write `g` and `h` for other densities, because some people are squeamish about overloading `f` for every density under discussion.

> 24.4:  "Standalone" looks odd to me.  I'd say "Stand-alone"

I can make it stand-alone, but both are OK according to the dictionary!

> In chapter 26, you might note that posterior predictive checks and prior predictive checks can be considered as two cases of the general concepts of predictive checks, just conditioning on different things.  For hierarchical models there are also intermediate versions.  See figure 1 of this paper: http://www.stat.columbia.edu/~gelman/research/published/A6n41.pdf
Perhaps unsurprisingly, the graphs in that figure are related to some of my thoughts involving workflow.  In any case, the key idea is that y^rep can be thought of as living within a graphical model that represents a kind of three-dimensional extension of the usual prior-likelihood-posterior space.  I think this is super-important!

I like the idea of unifying things and adding the graphs, but would rather get this PR in than try to keep enhancing.  It already got away from me---I was just going to do a PPC chapter!  We'll have to regenerate those graphs or get permission for includes.  I'd suggest also adding the SBC graphs that are similar.

avehtari commented on src/stan-users-guide/cross-validation.Rmd:
```diff
@@ -0,0 +1,390 @@
+when the test set is just a single item, the result is known as
+leave-one-out (LOO) cross-validation.
+
+Partitioning the data and reusing the partitions is very fiddly in the
+indexes and may not lead to even divisions of the data. It's far
+easier to use random partitions, which support arbitrarily sized
+test/training splits and can be easily implemented in Stan.
```
Maybe add also an example code for cross-validation with permutation given as data?

bob-carpenter replied:
Thanks for letting me know.  I added discussion.  

I'll just mention you can take the permutation in as data as easily as generating it by a function.  I'm not sure why you'd want to do that rather than just doing the test/training split on the outside if you want to control it.  

I was going to write the K-fold version, but it's hard to write the code in general when the data might not be divisible by K.  I could add it.

Ack---I also realize I completely failed to mention the loo package and the approach it takes to doing approximate LOO x-validation.  I'll do that, too, but with just a citation, not a lot of detail.  It'd probably be good to expand later to talk more about approximate LOO.

avehtari commented on src/stan-users-guide/decision-analysis.Rmd:
```diff
@@ -0,0 +1,250 @@
+
+## Outline of decision analysis
+
+Following [@GelmanEtAl:2013], Bayesian decision analysis can be
```
[@GelmanEtAl:2013] -> @GelmanEtAl:2013

bob-carpenter replied:
fixed.

bob-carpenter commented:
From Dave Blei:

> one small suggestion : the classical ppc is not well-calibrated.  see the references below.  (bayarri & berger's paper shows the applied side of this.)  i still think small p-values are cause for concern. but consider removing the clause about the p-value being uniform.

```
@article{Bayarri:2000,
Author = {Bayarri, MJ and Berger, James O},
Journal = {Journal of the American Statistical Association},
Number = {452},
Pages = {1127--1142},
Title = {P values for composite null models},
Volume = {95},
Year = {2000}}
```

avehtari commented:
> the classical ppc is not well-calibrated

True. I was thinking to suggest to remove ppc p-values as no-one (including Andrew) computes them in practice anymore due to this issue and as it unnecessarily brings hypothesis testing problems (e.g. dichotomization and arbitrary threshold). Unfortunately it's difficult to change this in every printed copy of BDA3. Non-calibration is directly related to ancillarity requirement I mentioned (which is also mentgioned in BDA3). You can easily test that ppc-p-values are not uniform with mean test statistic for that Poisson model. Posterior predictive checking is still useful quantitatively showing in which way the there is discrepancy  (and usually then p-value would 0 or 1, and plot shows what kind of problem there is).

bob-carpenter commented on src/stan-users-guide/posterior-prediction.Rmd:
```diff
@@ -0,0 +1,420 @@
+\log p(\tilde{y} \mid y)
+=
+- \log M
++ \textrm{log-sum-exp}_{m = 1}^M \log p(\tilde{y} \mid \theta^{(m)}).
```
Fixed with reference to expanded log-sum-exp chapter in the arithmetic precision chapter.

bob-carpenter commented on src/stan-users-guide/posterior-prediction.Rmd:
````diff
@@ -0,0 +1,420 @@
+quantity is computed by Stan if the value of $f(\tilde{y}, \theta)$ is assigned
+to a variable in the generated quantities block.  That is, if we have
+```
+generated quatnities {
````
Fixed.

bob-carpenter commented on src/stan-users-guide/simulation-based-calibration.Rmd:
Fixed, along with Cook-Gelman-Rubin in same sentence.

bob-carpenter commented on src/stan-users-guide/simulation-based-calibration.Rmd:
done.

bob-carpenter commented on src/stan-users-guide/posterior-predictive-checks.Rmd:
```diff
@@ -0,0 +1,381 @@
+size needed to ensure rough independence of the replications.
+
+Here's a complete example where the model is a simple Poisson with
+improper prior.
```
I added an exponential(0.1) weakly informative prior.  Good thing I had all the plotting scripted!

bob-carpenter commented on src/stan-users-guide/posterior-predictive-checks.Rmd:
```diff
@@ -0,0 +1,381 @@
+or a latent varying effects (aka random effects) model that can
+account for the over-dispersion.
+
+### Which statistics to test?
+
+Any statistic may be used for the data, but these can be guided by the
+quantities of interest in the model itself.  Popular choices in
+addition to mean and standard deviation are quantiles, such as the
+median, 5% or 95% quantiles, or even the maximum or minimum value to
+test extremes.
```
Thanks for pointing that out.  I hadn't thought about it that way before or I would've included that discussion.  I expanded the which-statistics-to-test section with a discussion of why the statistics shouldn't be directly derived from model parameters.  This is tricky, though.  The Poisson has one parameter for both mean and variance---we just happen to know that the posterior around the mean will fit better than the posterior around the variance.   We could derive that analytically, but I fear that's a complicated task in general.

bob-carpenter commented on src/stan-users-guide/posterior-predictive-checks.Rmd:
```diff
@@ -0,0 +1,381 @@
+
+The p-value for the mean is just the percentage of replicated data
+sets whose statistic is greater than or equal that of the original
+data.  Using a Poisson model for negative binomial data still fits the
+mean well, with a posterior $p$-value of 0.49.  In Stan terms, it is
+extracted as the posterior mean of the indicator variable `mean_gt`.
+
+The standard deviation statistic tells a different story.
```
Added in the relevant section.

bob-carpenter commented on src/stan-users-guide/cross-validation.Rmd:
```diff
@@ -0,0 +1,390 @@
+$$
+
+If the estimate is larger than the true value, the error is positive,
+and if oit's smaller, then error is negative.  If an estimator's
```
fixed.

bob-carpenter commented on src/stan-users-guide/cross-validation.Rmd:
```diff
@@ -0,0 +1,390 @@
+y^{\textrm{train}}).
+$$
+This just calculates error; taking absolute value or squaring will
+compute absolue error and mean square error. Note that the absolute
```
fixed.

bob-carpenter commented:
I addressed the issue with p-values not being calibrated in the text and cited Bayarri and Berger.

bob-carpenter commented on src/stan-users-guide/cross-validation.Rmd:
```diff
@@ -0,0 +1,390 @@
+
+If the estimate is larger than the true value, the error is positive,
+and if oit's smaller, then error is negative.  If an estimator's
+unbiased, then expected error is zero.  So typically, absolute error
+or squared error are used, which will always have positive
+expectations for an imperfect estimator.  *Absolute error* is defined as
```
Sure.  I really like that paper (and their paper on calibration).

bob-carpenter commented on src/stan-users-guide/simulation-based-calibration.Rmd:
done

bob-carpenter commented on src/stan-users-guide/simulation-based-calibration.Rmd:
done

bob-carpenter commented:
@avehtari : I addressed all your comments.  Thanks so much!  I addressed @andrewgelman's too, and added a section paralleling Gelman, Meng, and Stern's diagrams.  

I also made sure it built in pdf and html without glitches.

avehtari commented:
I checked the changes and all look good. log_lik computations in generated quantities for loo would be possible addition, but this is useful if merged as is.

bob-carpenter commented:
Thanks, @avehtari.  If you want to write that section, I'm happy to review.  I'd like it to explain how to do loo, not just point at the loo package, though.  This is for the general Stan doc.
